test/e2e/common/expansion.go

```diff
@@ -306,6 +307,137 @@ var _ = framework.KubeDescribe("Variable Expansion", func() {
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "tail -f /dev/null"},
+						Env: []v1.EnvVar{
+							{
+								Name:  "POD_NAME",
+								Value: "foo",
+							},
+							{
+								Name: "ANNOTATION",
+								ValueFrom: &v1.EnvVarSource{
+									FieldRef: &v1.ObjectFieldSelector{
+										APIVersion: "v1",
+										FieldPath:  "metadata.annotations['mysubpath']",
+									},
+								},
+							},
+						},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:        "workdir1",
+								MountPath:   "/logscontainer",
+								SubPathExpr: "$(ANNOTATION)/$(POD_NAME)",
+							},
+							{
+								Name:      "workdir2",
+								MountPath: "/testcontainer",
+							},
+						},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+				Volumes: []v1.Volume{
+					{
+						Name: "workdir1",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: "/tmp"},
+						},
+					},
+					{
+						Name: "workdir2",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: "/tmp"},
+						},
+					},
+				},
+			},
+		}
+
+		By("creating the pod with failed condition")
+		pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+		Expect(err).ToNot(HaveOccurred(), "while creating pod")
+
+		err = framework.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+		Expect(err).To(HaveOccurred(), "while waiting for pod to be running")
+
+		var podClient *framework.PodClient
+		podClient = f.PodClient()
+
+		By("updating the pod")
+		podClient.Update(podName, func(pod *v1.Pod) {
+			pod.ObjectMeta.Annotations = map[string]string{"mysubpath": "mypath"}
+		})
+
+		By("waiting for pod running")
+		err = framework.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+		Expect(err).NotTo(HaveOccurred(), "while waiting for pod to be running")
+
+		By("creating a file in subpath")
+		cmd := "touch /testcontainer/mypath/foo/test.log"
+		_, err = framework.RunHostCmd(pod.Namespace, pod.Name, cmd)
+		if err != nil {
+			framework.Failf("expected to be able to write to subpath")
+		}
+
+		By("test for file in mounted path")
+		cmd = "test -f /logscontainer/test.log"
+		_, err = framework.RunHostCmd(pod.Namespace, pod.Name, cmd)
+		if err != nil {
+			framework.Failf("expected to be able to verify file")
+		}
+
+		By("updating the annotation value")
+		podClient.Update(podName, func(pod *v1.Pod) {
+			pod.ObjectMeta.Annotations["mysubpath"] = "mynewpath"
+		})
+
+		By("waiting for annotated pod running")
+		err = framework.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+		Expect(err).NotTo(HaveOccurred(), "while waiting for annotated pod to be running")
+
+		By("test for subpath preserved in mounted path")
+		cmd = "test -f /testcontainer/mypath/foo/test.log"
```

Comment:
This should use the path from the subpath mount.  We want to validate that the subpath mount didn't change between container restarts.

Reply:
Added both mounts into the test for clarity